shared/js/dataDriver.js

```diff
@@ -305,15 +316,27 @@ loop.DataDriver = function() {
      *  - {Mixed}  value     Stored value for the record.
      */
     _processRecord(key, data) {
+      let dispatchAction, dispatchObject;
       // XXX akita bug 1274130: Validate data/key/timestamp before processing.
-      let [, type] = key.match(/^([^!]+)!(.+)$/);
+      let [, type, id] = key.match(/^([^!]+)!(.+)$/);
       switch (type) {
-        case "chat": {
-          let chat = data.value;
-          chat.receivedTimestamp = new Date(data.timestamp).toISOString();
-          this._dispatcher.dispatch(new actions.ReceivedTextChatMessage(chat));
+        case "chat":
+          dispatchAction = "ReceivedTextChatMessage";
+          dispatchObject = Object.assign(data.value, {
+            receivedTimestamp: new Date(data.timestamp).toISOString()
+          });
```

Comment:
Avoided modifying `data` param by refactoring to have a `dispatchExtra` object and combined them with `.assign` that starts with `{}` then adds `data.value` then adds `dispatchExtra`.